### PR TITLE
[feature/cumulative--10.0.0] AB#33495 - ActionBar Gap Closures

### PR DESCRIPTION
**Pull request by morethanfire**

### CODE REVIEW

**Comment by groberts314** [comment]
PR Review Checkpoint Meeting Thurs 2020-09-10.
Geoffrey to review today.

File: src/surfaces/actionBar/actionBar.jsx [hunk 1]
```diff
@@ -1,63 +1,49 @@
...
-        md: PropTypes.oneOf(GRID_SIZES),
-        search: PropTypes.shape({}),
-        sm: PropTypes.oneOf(GRID_SIZES),
-        xl: PropTypes.oneOf(GRID_SIZES),
-    })),
+    /**
+     * Array of objects that are used to setup
+     * the grid system in the ActionBar.
+     */
+    columns: COLUMNS_PROP_TYPES,
```
Comment (groberts314) [inline src/surfaces/actionBar/actionBar.jsx:43]: Good call defining a constant prop type for this! 👍

File: src/surfaces/actionBar/actionBar.jsx [hunk 2]
```diff
@@ -71,108 +57,124 @@ const propTypes = {
...
     justifyContent: 'flex-start',
-    moduleType: undefined,
-    style: {},
+    moduleType: null,
     toggleSmSearchVisibleClassName: null,
 };
 
 const styles = (theme) => {
+    const actionsButtonDrawerOptionHeight = 55;
     const standardThirdTiming = 300;
```
Comment (groberts314) [inline src/surfaces/actionBar/actionBar.jsx:85]: Should this be 333? 😁 
I kid ... if 300 works, it's all good.

File: src/surfaces/actionBar/actionBar.jsx [hunk 3]
```diff
@@ -71,108 +57,124 @@ const propTypes = {
...
-    const backwardsStep4Delay = `${backwardsStep3Delay + backwardsStep3Duration}ms`;
-    const backwardsStep4Duration = `${standardFifthTiming}ms`;
+    const backwardsStep4Delay = backwardsStep3Delay + backwardsStep3Duration;
+    const backwardsStep4Duration = standardFifthTiming;
 
     let optionContainerStyles;
     let subOptionStyles;
 
     for (let i = 1; i <= 11; i += 1) {
+        const optionYPos = i === 1 ? 0 : actionsButtonDrawerOptionHeight * (i - 1);
```
Comment (groberts314) [inline src/surfaces/actionBar/actionBar.jsx:143]: Is the ternary necessary here?  `i (===1) - 1 => 0 * actionsButtonDrawerOptionHeight  => 0` ... right?

File: src/surfaces/actionBar/actionBar.jsx [hunk 4]
```diff
@@ -489,314 +459,60 @@ class ActionBar extends React.Component {
...
-                                                {jsx}
-                                            </div>
-                                        </Grid.Column>
-                                    );
-                                })}
+                                <ActionBarGridColumns
+                                    columns={columns}
+                                    isMobileSearchVisible={isMobileSearchVisible}
+                                    onMobileSearchIconToggle={this.onMobileSearchIconToggle}
+                                />
```
Comment (groberts314) [inline src/surfaces/actionBar/actionBar.jsx:502]: Nice ... I like extracting this out into another component.  _Much_ cleaner. 👍

File: src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx [hunk 5]
```diff
@@ -45,23 +50,24 @@ class ActionBarActionsButtonDrawerOption extends React.PureComponent {
...
             option.onClick();
-            return false;
+
+            return null;
         }
 
         onClick(isSelected ? {} : option);
-        return true;
+
+        return null;
```
Comment (groberts314) [inline src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx:70]: I think we can just do early `return;` as necessary and elide final `return null;`.  I don't think we have to return a value.

I think the only exception to that is when rendering: we should _always_ return _something_ from rendering functions, i.e. `null` if we do in fact want to render nothing.  But for event handling functions and such, I think it is okay if they have void return type.  If your opinion differs, we can discuss.

File: src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx [hunk 6]
```diff
@@ -79,13 +86,13 @@ class ActionBarActionsButtonDrawerOption extends React.PureComponent {
             `${hide ? `actions_button_drawer--option_container-${idNumber}-hide` : ''}`,
             `${!hide && !isSelected ? `actions_button_drawer--option_container-${idNumber}-show` : ''}`,
             {
-                'actions_button_drawer--option_container-selected': isSelected,
+                [`actions_button_drawer--option_container-${idNumber}-selected`]: isSelected,
```
Comment (groberts314) [inline src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx:89]: If `idNumber` is `null` is it okay if these end up as `actions_button_drawer--option_container--show` and `actions_button_drawer--option_container--selected`?

File: src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx [hunk 7]
```diff
@@ -39,56 +40,52 @@ class ActionBarActionsButtonDrawerSubOption extends React.PureComponent {
...
+
+            return null;
         }
 
         if (isFunction(subOptionOnClick)) {
             subOptionOnClick();
         }
 
-        return true;
+        return null;
```
Comment (groberts314) [inline src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx:76]: Same deal with returning `null` versus void return type from `onClick`

File: src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx [hunk 8]
```diff
@@ -97,22 +94,27 @@ class ActionBarActionsButtonDrawerSubOption extends React.PureComponent {
...
 
-        /* eslint-disable jsx-a11y/click-events-have-key-events */
-        /* eslint-disable jsx-a11y/interactive-supports-focus */
         return (
             <div
                 className={subOptionClasses}
                 onClick={this.onClick}
+                onKeyDown={this.onKeyDown}
                 role="menuitem"
+                tabIndex={0}
```
Comment (groberts314) [inline src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx:117]: Just confirming: `tabIndex` and `onKeyDown` completely appease the `jsx-a11y` ESLINT warnings?

**Review by groberts314 — COMMENTED** [review]
Looking pretty sharp!  I had a few questions and suggestions, but I don't think much of it is blocking.

Additional Question: What's with the empty container and content files in docs?
`docs/src/layout/container/container.jsx`
`docs/src/layout/content/content.jsx`

Are these placeholders for future TODOs?

File: docs/src/global/example.jsx [hunk 9]
```diff
@@ -89,7 +98,14 @@ function Example(props) {
...
             >
-                {children}
+                {iframe ? (
+                    <ExampleFrame
+                        title={title}
+                        maxWidth={maxWidth}
+                    >
+                        {children}
+                    </ExampleFrame>
+                ) : children}
```
[the remark this answers is not in the document]
Comment (morethanfire) [reply docs/src/global/example.jsx:108]: Really the only need so far is for using the Page components.

File: docs/src/surfaces/actionBar/actionBar.jsx [hunk 10]
```diff
@@ -46,24 +63,89 @@ function DocsActionBar() {
...
+                >
+                    <ActionBarSmSample />
+                </Example>
+
+                <MarkdownContainer>
+                    <Typography
+                        variant="body1"
+                    >
+                        {/* eslint-disable-next-line react/jsx-one-expression-per-line */}
+                        Medium (<code>md</code>) breakpoint.
```
[the remark this answers is not in the document]
Comment (morethanfire) [reply docs/src/surfaces/actionBar/actionBar.jsx:85]: I totally agree.

File: docs/src/global/exampleFramed.jsx [hunk 11]
```diff
@@ -0,0 +1,76 @@
+import { create } from 'jss';
+import jssPreset from 'react-cm-ui/styles/jssPreset';
+import PropTypes from 'prop-types';
+import React, {
+    useCallback,
+    useEffect,
+    useMemo,
```
[the remark this answers is not in the document]
Comment (morethanfire) [reply docs/src/global/exampleFramed.jsx:7]: Yeah, it returns a memoized value.

[the remark this answers is not in the document]
Comment (morethanfire) [reply docs/src/global/exampleFramed.jsx:7]: https://reactjs.org/docs/hooks-reference.html#usememo

File: src/inputs/input/input.jsx [hunk 12]
```diff
@@ -60,10 +74,10 @@ const propTypes = {
...
     showSpinners: PropTypes.bool,
     style: PropTypes.shape({}),
-    tabIndex: PropTypes.oneOfType([
-        PropTypes.number,
-        PropTypes.string,
-    ]),
+    /**
+     * Indicates whether or not the Input can be focused.
+     */
+    tabIndex: PropTypes.number,
```
[the remark this answers is not in the document]
Comment (morethanfire) [reply src/inputs/input/input.jsx:80]: I need Rachel! 

I'll fix it up. thanks!

File: src/surfaces/actionBar/__test__/actionBarGridColumns.test.js [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+        expect(wrapper.find(`.${iconFilterClass}`).exists()).toEqual(true);
+        expect(wrapper.find(`.${iconGridClass}`).exists()).toEqual(true);
+        expect(wrapper.find(`.${iconSearchClass}`).exists()).toEqual(true);
+        expect(wrapper.find(`.${iconSettingsClass}`).exists()).toEqual(true);
+        expect(wrapper.find(`.${iconTableClass}`).exists()).toEqual(true);
+        expect(wrapper.find('ActionBarSearch').exists()).toEqual(true);
+        expect(wrapper.find(`.${jsxClass}`).exists()).toEqual(true);
+    });
+
+    it('Should foo', () => {
```
[the remark this answers is not in the document]
Comment (morethanfire) [reply src/surfaces/actionBar/__test__/actionBarGridColumns.test.js:130]: Nope, that's all you get. Bwahaha. Whoops. I'll get it fixed up.

File: src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx (hunk 8 above)
Comment (morethanfire) [reply src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx:117]: Yes, and also it's helpful to have `0` as the value so we can actually tab these actions and also hit enter/return. I am trying to focus on some keyboard things in 10.0.0.... trying.

File: docs/src/app/navigationConstants.js [hunk 14]
```diff
@@ -264,11 +264,6 @@ export const navigationItems = [
                 directory: 'layout',
                 label: 'Layout',
                 levelThree: [
-                    {
-                        component: 'containerDeprecated',
-                        label: 'Container (deprecated)',
-                        path: 'container-deprecated',
-                    },
```
[the remark this answers is not in the document]
Comment (morethanfire) [reply docs/src/app/navigationConstants.js:271]: Ha!

File: src/inputs/phoneInput/phoneInputCountryDropdownValue.jsx [hunk 15]
```diff
@@ -2,6 +2,10 @@ import ClassNames from 'classnames';
 import React from 'react';
 
 function PhoneInputCountryDropdownValue(selectedOption) {
+    if (!selectedOption) {
+        return null;
+    }
+
     const containerClasses = ClassNames('Select-value');
 
     if (!selectedOption) {
```
[the remark this answers is not in the document]
Comment (morethanfire) [reply src/inputs/phoneInput/phoneInputCountryDropdownValue.jsx]: Fixed.

File: src/inputs/prompt/prompt.jsx [hunk 16]
```diff
@@ -76,15 +76,17 @@ class Prompt extends React.Component {
...
+        }
 
         if (!_.isUndefined(onClick)) {
             onClick(option);
         } else {
             this.setState({ show: true });
         }
 
-        return false;
+        return null;
```
[the remark this answers is not in the document]
Comment (morethanfire) [reply src/inputs/prompt/prompt.jsx:89]: fixed.

File: src/surfaces/actionBar/__test__/actionBarGridColumns.test.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (morethanfire) [reply src/surfaces/actionBar/__test__/actionBarGridColumns.test.js]: Fixed.

File: src/surfaces/actionBar/actionBar.jsx (hunk 2 above)
Comment (morethanfire) [reply src/surfaces/actionBar/actionBar.jsx:85]: Probably, but I don't think I should change it because of it potentially messing up the animation.

Comment (morethanfire) [reply src/surfaces/actionBar/actionBar.jsx]: Nm, I fixed it.

File: src/inputs/phoneInput/phoneInputCountryDropdownValue.jsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (groberts314) [reply src/inputs/phoneInput/phoneInputCountryDropdownValue.jsx]: Looks good.  I guess maybe this was some merge issue or some such, eh?  In any event, resolved.

File: src/inputs/prompt/prompt.jsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (groberts314) [reply src/inputs/prompt/prompt.jsx]: Looks good.  Resolving.

File: src/surfaces/actionBar/__test__/actionBarGridColumns.test.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (groberts314) [reply src/surfaces/actionBar/__test__/actionBarGridColumns.test.js]: Looks good.  Resolving,

File: src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx (hunk 5 above)
Comment (morethanfire) [reply src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx:70]: Fixed.

File: src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx (hunk 7 above)
Comment (morethanfire) [reply src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx:76]: Fixed.

File: src/surfaces/actionBar/actionBar.jsx (hunk 3 above)
Comment (morethanfire) [reply src/surfaces/actionBar/actionBar.jsx:141]: Nope it isn't. Thanks. Fixed.

File: src/surfaces/actionBar/actionBar.jsx (hunk 2 above)
Comment (groberts314) [reply src/surfaces/actionBar/actionBar.jsx]: Looks like we're now using "standard" transition durations defined by Material UI eh?  Me likey.
Resolving.

File: src/surfaces/actionBar/actionBar.jsx (hunk 3 above)
Comment (groberts314) [reply src/surfaces/actionBar/actionBar.jsx]: Looks simplified now.  Thanks.  Resolving this.

File: src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx (hunk 6 above)
Comment (morethanfire) [reply src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx:87]: Fixed. Default value is `1` now.

File: src/inputs/input/input.jsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (morethanfire) [reply src/inputs/input/input.jsx:80]: Fixed.

File: src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx (hunk 5 above)
Comment (groberts314) [reply src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx]: Looks good now (void returns).  Resolving.

**Comment by morethanfire** [comment]
@groberts314, I addressed all your concerns. Thanks you for the review!

File: src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx (hunk 6 above)
Comment (groberts314) [reply src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx:87]: Sounds good.
Also should `option_container-${idNumber}` be `option_container_${idNumber}`?
Like, is all of that the "element", and `selected` is the "modifier"?

File: src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx (hunk 7 above)
Comment (groberts314) [reply src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx]: Looks good.  Resolving.

File: src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx (hunk 8 above)
Comment (groberts314) [reply src/surfaces/actionBar/actionBarActionsButtonDrawerSubOption.jsx:111]: Copy that.  Carry on, good sir!
Resolving.

File: src/inputs/input/input.jsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (groberts314) [reply src/inputs/input/input.jsx]: Oh nice!  Love the MDN link too!
Resolving.

**Comment by groberts314** [comment]
@morethanfire One more small thing I noticed: https://github.com/saddlebackdev/react-cm-ui/pull/202#discussion_r493914516

File: src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx (hunk 6 above)
Comment (morethanfire) [reply src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx:87]: No, the id and `selected` are the modifiers.

Comment (groberts314) [reply src/surfaces/actionBar/actionBarActionsButtonDrawerOption.jsx:87]: Very well.  I'll buy what you're selling.  We can resolve this thread if you like.

**Review by groberts314 — APPROVED** [review]
Looks good to me,

**Pull request merged**